File: _scene/FXDsceneTable.m


#import "FXDsceneTable.h"


@implementation FXDsceneTable	

#pragma mark - Memory management

#pragma mark - Initialization

#pragma mark - Autorotating

#pragma mark - View Appearing
- (void)willMoveToParentViewController:(UIViewController *)parent {

	if (parent == nil) {
		self.mainTableview.delegate = nil;
		self.mainTableview.dataSource = nil;
	}

	[super willMoveToParentViewController:parent];
}


#pragma mark - Property overriding
- (UIScrollView*)mainScrollview {
	if (_mainScrollview == nil) {
		
		if (self.mainTableview) {
			_mainScrollview = self.mainTableview;
		}
		else {
			_mainScrollview = [super mainScrollview];
		}
	}
	
	return _mainScrollview;
}


#pragma mark - Method overriding
- (void)registerMainCellNib {
	
	if ((self.mainCellNib == nil && self.mainCellIdentifier == nil)
		|| self.mainTableview == nil) {
		[super registerMainCellNib];
		return;
	}


	FXDLog_DEFAULT;
	FXDLogObject(self.mainCellNib);
	FXDLogObject(self.mainCellIdentifier);

	FXDLogObject(self.mainTableview);
	[self.mainTableview registerNib:self.mainCellNib forCellReuseIdentifier:self.mainCellIdentifier];
}


#pragma mark - IBActions

#pragma mark - Public
- (void)initializeTableCell:(FXDTableViewCell*)cell forIndexPath:(NSIndexPath*)indexPath {
	//NOTE: for newly initialized cell.
}

- (void)configureTableCell:(FXDTableViewCell*)cell forIndexPath:(NSIndexPath*)indexPath {
#if ForDEVELOPER
	if (self.mainCellNib) {
		FXDLog(@"WARNING: %@", @"Make sure to configure properly");
	}
#endif

	[self configureSectionPostionTypeForTableCell:cell forIndexPath:indexPath];
	
	UIImage *backgroundImage = [self backgroundImageForTableCellAtIndexPath:indexPath];
	UIImage *highlightedImage = [self selectedBackgroundImageForTableCellAtIndexPath:indexPath];
	[cell customizeBackgroundWithImage:backgroundImage withHighlightedImage:highlightedImage];
	
	cell.textLabel.text = (self.cellTitleDictionary)[indexPath];
	cell.detailTextLabel.text = (self.cellSubTitleDictionary)[indexPath];
	
	UIImage *mainImage = [self mainImageForTableCellAtIndexPath:indexPath];
	UIImage *highlightedMainImage = [self highlightedMainImageForTableCellAtIndexPath:indexPath];
	[cell customizeWithMainImage:mainImage withHighlightedMainImage:highlightedMainImage];
	
	cell.accessoryView = [self accessoryViewForTableCellAtIndexPath:indexPath];
}

- (void)configureSectionPostionTypeForTableCell:(FXDTableViewCell*)cell forIndexPath:(NSIndexPath*)indexPath {

	NSInteger rowCount = [(self.itemCounts)[indexPath.section] integerValue];

	if (rowCount == 1) {
		cell.sectionPositionType = sectionPositionOne;
	}
	else if (rowCount > 1) {
		if (indexPath.row == 0) {
			cell.sectionPositionType = sectionPositionTop;
		}
		else if (indexPath.row == rowCount-1) {
			cell.sectionPositionType = sectionPositionBottom;
		}
		else {
			cell.sectionPositionType = sectionPositionMiddle;
		}
	}
}

- (UIImage*)backgroundImageForTableCellAtIndexPath:(NSIndexPath*)indexPath {
	UIImage *backgroundImage = nil;
	
	return backgroundImage;
}

- (UIImage*)selectedBackgroundImageForTableCellAtIndexPath:(NSIndexPath*)indexPath {
	UIImage *backgroundImage = nil;
	
	return backgroundImage;
}

#pragma mark -
- (UIImage*)mainImageForTableCellAtIndexPath:(NSIndexPath*)indexPath {
	UIImage *mainImage = nil;

	return mainImage;
}

- (UIImage*)highlightedMainImageForTableCellAtIndexPath:(NSIndexPath*)indexPath {
	UIImage *highlightedMainImage = nil;

	return highlightedMainImage;
}

- (UIView*)accessoryViewForTableCellAtIndexPath:(NSIndexPath*)indexPath {
	id accessoryView = nil;

	if (accessoryView) {
		if ([(UIImageView*)accessoryView image] == nil) {
			accessoryView = nil;
		}
	}
	
	return (UIView*)accessoryView;
}

- (UIView*)sectionDividerViewForWidth:(CGFloat)width andHeight:(CGFloat)height {
	CGRect dividerFrame = CGRectMake(0.0, 0.0, width, height);
	
	UIView *sectionDividingView = [[UIView alloc] initWithFrame:dividerFrame];
	sectionDividingView.backgroundColor = [UIColor clearColor];
	
	return sectionDividingView;
}

#pragma mark - Observer

#pragma mark - Delegate
//NOTE: UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
	NSInteger numberOfSections = [self numberOfSectionsForScrollView:tableView];
	
	return numberOfSections;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
	NSInteger numberOfRows = [self numberOfItemsForScrollView:tableView atSection:section];
	
	return numberOfRows;
}

#pragma mark -
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
	
	FXDTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:self.mainCellIdentifier];

	if (cell == nil) {
		cell = [[FXDTableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:self.mainCellIdentifier];

		[self initializeTableCell:cell forIndexPath:indexPath];
	}

	[self configureTableCell:cell forIndexPath:indexPath];
	
	
	return cell;
}


//NOTE: UITableViewDelegate
- (void)tableView:(UITableView *)tableView didEndDisplayingCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath*)indexPath {
	
	BOOL didCancel = [self.cellOperationQueue
					  cancelOperationForKey:indexPath
					  withDictionary:self.cellOperationDictionary];

	if (didCancel) {
		//FXDLog(@"%@ %@", _BOOL(didCancel), _Object(indexPath));
	}
}

@end

File: fXDKit/Classes/FXDsceneTable.swift


public protocol FXDsceneWithCells {
	var mainCellIdentifier: String { get }
	var mainDataSource: NSMutableArray? { get set }

	var cellOperationQueue: OperationQueue? { get }
	var cellOperationDictionary: NSMutableDictionary? { get }
}

extension FXDsceneWithCells {
	public var mainCellIdentifier: String {
		fxd_overridable()
		return "CELL_\(String(describing: type(of: self)))"
	}
}


public protocol FXDsceneWithTableCells {
	var cellTitleDictionary: [String : String] { get }
	var cellSubTitleDictionary: [String : String] { get }

	func initializeTableCell(_ cell: UITableViewCell?, for indexPath: IndexPath!)
	func configureTableCell(_ cell: UITableViewCell?, for indexPath: IndexPath!)
	func configureSectionPosition(forTableCell cell: UITableViewCell?, for indexPath: IndexPath!)

	func backgroundImageForTableCell(at indexPath: IndexPath!) -> UIImage?
	func selectedBackgroundImageForTableCell(at indexPath: IndexPath!) -> UIImage?
	func mainImageForTableCell(at indexPath: IndexPath!) -> UIImage?
	func highlightedMainImageForTableCell(at indexPath: IndexPath!) -> UIImage?
	func accessoryViewForTableCell(at indexPath: IndexPath!) -> UIView?
	func sectionDividerView(forWidth width: CGFloat, andHeight height: CGFloat) -> UIView?
}


open class FXDsceneTable: UIViewController, FXDsceneScrollable, FXDsceneWithCells, FXDsceneWithTableCells {
	@IBOutlet open var mainTableview: UITableView?


	open var mainScrollview: UIScrollView? {
		return mainTableview
	}

	open var mainDataSource: NSMutableArray?


	open lazy var cellOperationQueue: OperationQueue? = {
		return OperationQueue.newSerialQueue(withName: String(describing: self))
	}()
	open lazy var cellOperationDictionary: NSMutableDictionary? = {
		return NSMutableDictionary.init()
	}()


	open var cellTitleDictionary: [String : String] {
		return [:]
	}

	open var cellSubTitleDictionary: [String : String] {
		return [:]
	}

	open func initializeTableCell(_ cell: UITableViewCell?, for indexPath: IndexPath!) {
		fxd_overridable()
	}

	open func configureTableCell(_ cell: UITableViewCell?, for indexPath: IndexPath!) {
		cell?.textLabel?.text = cellTitleDictionary[indexPath.stringKey]
		cell?.detailTextLabel?.text = cellSubTitleDictionary[indexPath.stringKey]
		cell?.accessoryView = accessoryViewForTableCell(at: indexPath)

		if let fxdCell = cell as? FXDTableViewCell {
			let backgroundImage = backgroundImageForTableCell(at: indexPath)
			let highlightedImage = selectedBackgroundImageForTableCell(at: indexPath)
			fxdCell.customizeBackground(with: backgroundImage, withHighlightedImage: highlightedImage)

			let mainImage = mainImageForTableCell(at: indexPath)
			let highlightedMainImage = highlightedMainImageForTableCell(at: indexPath)
			fxdCell.customize(withMainImage: mainImage, withHighlightedMainImage: highlightedMainImage)
		}
	}

	open func configureSectionPosition(forTableCell cell: UITableViewCell?, for indexPath: IndexPath!) {
		guard let fxdCell = cell as? FXDTableViewCell else {
			return
		}

		let itemCount = tableView(mainTableview!, numberOfRowsInSection: indexPath.section)

		if (itemCount == 1) {
			fxdCell.positionCase = .single
		}
		else if (itemCount > 1) {
			if (indexPath.row == 0) {
				fxdCell.positionCase = .top
			}
			else if (indexPath.row == itemCount-1) {
				fxdCell.positionCase = .bottom
			}
			else {
				fxdCell.positionCase = .middle
			}
		}
	}


	open func backgroundImageForTableCell(at indexPath: IndexPath!) -> UIImage? {
		return nil
	}

	open func selectedBackgroundImageForTableCell(at indexPath: IndexPath!) -> UIImage? {
		return nil
	}

	open func mainImageForTableCell(at indexPath: IndexPath!) -> UIImage? {
		return nil
	}

	open func highlightedMainImageForTableCell(at indexPath: IndexPath!) -> UIImage? {
		return nil
	}

	open func accessoryViewForTableCell(at indexPath: IndexPath!) -> UIView? {
		return nil
	}

	open func sectionDividerView(forWidth width: CGFloat, andHeight height: CGFloat) -> UIView? {
		let dividerFrame = CGRect(x: 0.0, y: 0.0, width: width, height: height)

		let sectionDividingView = UIView(frame: dividerFrame)
		sectionDividingView.backgroundColor = .clear

		return sectionDividingView
	}
}

extension FXDsceneTable {
	override open func willMove(toParentViewController parent: UIViewController?) {
		if parent == nil {
			cellOperationQueue?.resetOperationQueueAndDictionary(cellOperationDictionary)
		}

		super.willMove(toParentViewController: parent)
	}
}


extension FXDsceneTable: UITableViewDataSource {
	open func numberOfSections(in tableView: UITableView) -> Int {
		return 1	//MARK: Assume it's often just one array
	}

	open func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return (mainDataSource != nil) ? (mainDataSource?.count)! : 0
	}

	open func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		var cell = tableView.dequeueReusableCell(withIdentifier: mainCellIdentifier)

		if cell == nil {
			cell = FXDTableViewCell.init(style: .subtitle, reuseIdentifier: mainCellIdentifier)

			initializeTableCell(cell, for: indexPath)
		}

		configureTableCell(cell, for: indexPath)

		return cell!
	}
}

extension FXDsceneTable: UITableViewDelegate {
	open func tableView(_ tableView: UITableView, didEndDisplaying cell: UITableViewCell, forRowAt indexPath: IndexPath) {
		_ = cellOperationQueue?.cancelOperation(forKey: indexPath, with: cellOperationDictionary)
	}
}
